src/Behat/Testwork/ServiceContainer/ExtensionManager.php

```diff
@@ -34,30 +32,46 @@
     private $debugInformation = array(
         'extensions_list' => array()
     );
+    /**
+     * @var ExtensionInstantiator[]
+     */
+    private $instantiators = array();
 
     /**
      * Initializes manager.
      *
-     * @param Extension[] $extensions     List of default extensions
-     * @param null|string $extensionsPath Base path where to search custom extension files
+     * @param Extension[]             $extensions      List of default extensions
+     * @param string|null             $extensionsPath  Base path where to search custom extension files (deprecated)
+     * @param ExtensionInstantiator[] $instantiators   Extension instantiators to use
      */
-    public function __construct(array $extensions, $extensionsPath = null)
+    public function __construct(array $extensions, $extensionsPath = null, array $instantiators = array())
     {
         foreach ($extensions as $extension) {
             $this->extensions[$extension->getConfigKey()] = $extension;
         }
 
-        $this->extensionsPath = $extensionsPath;
+        // BC Layer, using default instantiators
+        // to be removed in 4.0
+        $this->instantiators = empty($instantiators)
+            ? array(new ExtensionClassInstantiator(), new ExtensionFileInstantiator($extensionsPath))
+            : $instantiators
+        ;
     }
 
     /**
      * Sets path to directory in which manager will try to find extension files.
      *
      * @param null|string $path
+     * @deprecated since 3.4, to be removed in 4.0
```

Comment:
I suggest adding deprecation warnings using the Symfony convention (i.e. `@trigger_error('...', E_USER_DEPRECATED);`) in deprecated APIs.

Behat does not provide reporting of usage of deprecated APIs yet (I think there is a third-party extension doing it though), but it may still help people using  Behat programmatically (which are the only ones caring about it, as the built-in usage of Behat should not trigger deprecated APIs for the ExtensionManager)